### PR TITLE
Dev 3

### PR DESCRIPTION
**Pull request by Tarikyil1998**
Please review my beautifull code which I put my blood, sweat and bodyjuices in.  

Thanks.

### CODE REVIEW

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 1]
```diff
@@ -0,0 +1,491 @@
+﻿using LibNoise.Generator;
+using System.Collections.Generic;
+using System.Text;
+using UnityEngine;
+
+public class FractalTree : MonoBehaviour
+{
+
+    public Dictionary<char, string> rules = new Dictionary<char, string>();
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:9]: Public variables start with a capital letter, please look at all of the added code and change this

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+    [Range(0, 4)]
+    public int iterations;
+    public string input = "F";
+    private string output;
+
+    public string result;
+
+    Perlin perlin;
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:17]: Every variable should have an access modifier, please look at all of the added code and change this

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+    void Start()
+    {
+        float frequency = 1f;
+        float persistence = 1f;
+        float lacunarity = 2f;
+        int octaves = 1;
+        perlin = new Perlin(frequency, lacunarity, persistence, octaves, GameObject.Find("Level").GetComponent<TerrainGenerator>().Seed, LibNoise.QualityMode.High);
+    }
+
+    public GameObject GenerateTree(Vector3 startPosition , BiomeType biome)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:54]: Remove extra space between startPosition and ,

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+    {
+        (treeMaterial,leafPrefab,leafMaterial,plantColor) = getBiomePrefabs(biome);
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:56]: Add spaces between variables

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+        if (color)
+        {
+            gameObject.GetComponent<MeshRenderer>().material.color = plantColor;
+        }
+        gameObject.transform.GetComponent<MeshFilter>().mesh = new Mesh();
+        gameObject.transform.GetComponent<MeshFilter>().mesh.indexFormat = UnityEngine.Rendering.IndexFormat.UInt32;
+        gameObject.transform.GetComponent<MeshFilter>().mesh.CombineMeshes(combine);
+        gameObject.transform.gameObject.SetActive(true);
+    }
+    string applyRules(string p_input)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:159]: Add whitespace and access modifier, also rename p_input to input, we dont use this naming convention

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+    // Create Point with a Position, Angle and BranchLength
+    struct point
+    {
+        public point(Vector3 rP, Vector3 rA, float rL) { Point = rP; Angle = rA; BranchLength = rL; }
+        public Vector3 Point;
+        public Vector3 Angle;
+        public float BranchLength;
+    }
+    // Determine Points using a String input
+    void determinePointsTree(string p_input)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:188]: Add access modifier

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+        point lastPoint = new point(Vector3.zero, Vector3.zero, 0.7f); 
+        returnValues.Push(lastPoint);
+
+        foreach (char c in p_input)
+        {
+            double sampleX = (newPosition.x) / scale;
+            double sampleY = (newPosition.y) / scale;
+
+            float perlinValue = (float)(perlin.GetValue(sampleY + lastPoint.Angle.y, 0, sampleX + lastPoint.Angle.y) + 1) / 2 * 100;
+            //perlinValue = perlinValue * 10f;
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:202]: Remove commented out code

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 8]
```diff
@@ -0,0 +1,491 @@
...
+                    break;
+                case ']': // Load Saved State
+                    lastPoint = returnValues.Pop();
+                    break;
+            }
+        }
+    }
+
+    // Create Cylinders using a startPositiong
+
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:291]: Remove whitespace between comment and function, startPositiong > startPosition

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+                case ']': // Load Saved State
+                    lastPoint = returnValues.Pop();
+                    break;
+            }
+        }
+    }
+
+    // Create Cylinders using a startPositiong
+   
+    void CreateCylinders(Vector3 startPosition)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:292]: Add access modifier

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 10]
```diff
@@ -0,0 +1,491 @@
...
+    {
+
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:294]: Remove whitespace

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+        float trunkSize = 0.3f;
+        float prevTrunkSize = 0.3f;
+        float branchSize = 1f;
+        float oldBranchSize = 10f;
+
+        for (int i = 0; i < points.Count; i += 2)
+        {
+            // Prevent Branchsize/oldBrancheSize from going under 0
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:302]: oldBrancheSize > oldBranchSize

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+            if (branchSize < 0f)
+            {
+                branchSize = 0.01f;
+            }
+            if (oldBranchSize < 0f)
+            {
+                oldBranchSize = 0.01f;
+            }
+            // If Start Point
+            if (points[i].Point.x == 0 && points[i + 1].Point.x == 0)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:312]: Should values like 0.003f, 3.6f, 0.1f, 0.01f be numbers or variables? If multiple uses of the same number represent the same thing, make them variables

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+                branchSize = oldBranchSize - 0.01f;
+
+                CreateCylinder(points[i], points[i + 1], 0.1f, startPosition, branchSize, oldBranchSize, false, false);
+                oldBranchSize = branchSize;
+                CreateLeaf(points[i], points[i + 1], branchSize, false);
+            }
+        }
+    }
+
+    // Create Plants using a startPoisition
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:339]: startPoisition > startPosition

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 14]
```diff
@@ -0,0 +1,491 @@
...
+            if (branchSize < 0f)
+            {
+                branchSize = 0.1f;
+            }
+            if (oldBranchSize < 0f)
+            {
+                oldBranchSize = 0.1f;
+            }
+            // Start Point
+            if (points[i].Point.x == 0 && points[i + 1].Point.x == 0)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:361]: Same comments as above, this function is also very similar to the one above, can they be combined?

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+                }
+                branchSize = oldBranchSize - 0.01f;
+
+                CreateCylinder(points[i], points[i + 1], 0.1f, startPosition, branchSize, oldBranchSize, false, true);
+                oldBranchSize = branchSize;
+                CreateLeaf(points[i], points[i + 1], branchSize, true);
+            }
+        }
+    }
+    // Pivot point1 around point2 by angles
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:388]: Add whitespace

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+    Vector3 pivot(Vector3 point1, Vector3 point2, Vector3 angles)
+    {
+        Vector3 dir = point1 - point2;
+        dir = Quaternion.Euler(angles) * dir;
+        point1 = dir + point2;
+        return point1;
+    }
+
+    // Cylinders are being placed between 2 points with variating sizes to create trunks/branches
+    void CreateCylinder(point point1, point point2, float radius, Vector3 startPosition, float oldSize, float newSize, bool lastTrunk, bool isPlant)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:398]: Add access modifier

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+            Vector3 scale = new Vector3(point1.BranchLength, length, point1.BranchLength - 0.01f);
+            newCylinder.transform.localScale = scale;
+            newCylinder.transform.position = newCylinder.transform.position + point1.Point;
+            newCylinder.transform.Rotate(point2.Angle);
+            branches.Add(newCylinder);
+            newCylinder.transform.parent = trunks.transform;
+        }
+        
+    }
+    // Create a Leaf based on Points position
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:428]: Add whitespace

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 18]
```diff
@@ -0,0 +1,491 @@
...
+    void CreateLeaf(point point1, point point2, float scale, bool isPlant)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:429]: Add access modifier

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs [hunk 19]
```diff
@@ -0,0 +1,491 @@
...
+            point2.Angle = Vector3.RotateTowards(point2.Angle, new Vector3(0, 0, 0), 1, 0.0f);
+            point2.Angle.y = point2.Angle.y + 180;
+            point2.Angle.x = -80;
+            leaf.transform.rotation = Quaternion.Euler(point2.Angle);
+            leaf.transform.parent = leaves.transform;
+        }
+    }
+
+    // Use Values from Editor to get different Tree,Leaves and Plants per biome
+    (Material TreeMaterial, GameObject LeavesPrefab, Material LeavesMaterial,Color plantColor) getBiomePrefabs(BiomeType biome)
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/Helpers/FractalTree.cs:456]: Add space after , and before Color, function name should start with a capital letter

File: No FYS-X Sky/Assets/World/Builders/WorldBuilder.cs [hunk 20]
```diff
@@ -16,6 +16,8 @@ class WorldBuilder
 
     private int regionRenderDistance;
 
+    private DecorationGenerator decorationGenerator;
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/World/Builders/WorldBuilder.cs:19]: This is unused, remove

File: No FYS-X Sky/Assets/World/Generators/DecorationGenerator.cs [hunk 21]
```diff
@@ -1,16 +1,70 @@
...
+
+    public float PlantThreshold;
+    public float TreeThreshold;
+    [Range(0.01f, 1)]
+    public float MarginThreshold;
+
+    [Range(1,3)]
+    public int interval;
+
+    Perlin perlin;
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/World/Generators/DecorationGenerator.cs:16]: Add access modifier

File: No FYS-X Sky/Assets/World/Tile.cs [hunk 22]
```diff
@@ -7,6 +7,7 @@ public class Tile
     public TileInfo Cave;
     public TileInfo Ocean;
     public List<GameObject> BuildingObjects;
+    public List<GameObject> DecoObjects;
```
Comment (Fakerick1) [inline No FYS-X Sky/Assets/World/Tile.cs:10]: Just call this DecorationObjects to prevent confusion

**Review by Fakerick1 — CHANGES_REQUESTED** [review]
Folder structure:

- _TerrainAutoUpgrade, Town Creator KIT Lite should be in the packages or helpers folder? Please review which of these are actually used
- Shaders 1, Textures 1, what even are these folders? Please move to Shaders and Textures, or in a subfolder of these folders if they must be separate (might be Stijns fault)
- Images folder? What are these used for (might also be Stijns fault)
- Ciconia Studio, what even is this? Remove the parts of whatever this is we dont use, if we dont use it at all, remove it.
- Scripts folder, what are these used for?

image

What are these files doing in the root folder

image

What is this file doing in the assets folder

File: No FYS-X Sky/Assets/Helpers/FractalTree.cs (hunk 14 above)
Comment (Tarikyil1998) [reply No FYS-X Sky/Assets/Helpers/FractalTree.cs:361]: They use different values If I combine them its gonna be a one big messy method thats why I seperated them.

**Pull request merged**